Add missing docs (#1481)
3scale/APIcast: gateway/src/apicast/policy/token_introspection/README.md

# OAuth 2.0 Token Introspection

>  The OAuth 2.0 Token Introspection policy allows validating the JSON Web Token (JWT) token used for services with the OpenID Connect (OIDC) authentication option using the Token Introspection Endpoint of the token issuer.

Reviewer: ```suggestion
 The OAuth 2.0 Token Introspection Policy allows validating the JSON Web Token (JWT) used for services with the OpenID Connect (OIDC) authentication option using the Token Introspection Endpoint of the token issuer.
```

Little typo.

APIcast supports the following authentication types in the `auth_type` field to determine the Token Introspection Endpoint and the credentials APIcast uses when calling this endpoint:

> * `use_3scale_oidc_issuer_endpoint`: APIcast uses the client credentials, Client ID and Client Secret, as well as the Token Introspection Endpoint from the OIDC Issuer setting configured on the Service Integration page. APIcast discovers the Token Introspection endpoint from the token_introspection_endpoint field. This field is located in the .well-known/openid-configuration endpoint that is returned by the OIDC issuer.

Reviewer: ```suggestion
* `use_3scale_oidc_issuer_endpoint`: APIcast uses the client credentials, Client ID, and Client Secret, as well as the Token Introspection Endpoint from the OIDC Issuer setting configured on the Service Integration page. APIcast discovers the Token Introspection Endpoint from the `token_introspection_endpoint` field. This field is located in the `.well-known/openid-configuration` endpoint that is returned by the OIDC issuer.
```

A couple of minor edits.

> * `client_id+client_secret`: specify a different Token Introspection Endpoint, as well as the Client ID and Client Secret APIcast uses to request token information.

Reviewer: @tkan145 This sentence sounds a bit confusing to me. I am not sure if it says:
It specifies a different Token Introspection Endpoint. As well as the Client ID and the Client Secret APIcast, it uses it to request token information.
or
Specify a different Token Introspection Endpoint, a different Client ID and a different Client Secret APIcast to request token information.

> * `client_secret_jwt`: Request token information using `client_secret_jwt` method. Prior to a token information request, APIcast will prepare a new JWT authentication token and sign using HMAC SHA-256 and with Client Secret as the shared key. APIcast will then make a token request including the generated client assertion as the value of the `client_assertion` parameter.

Reviewer: ```suggestion
* `client_secret_jwt`: It is a request token information using `client_secret_jwt` method. Prior to a token information request, APIcast will prepare a new JWT authentication token and sign-in using HMAC SHA-256 and the Client Secret as the shared key. Then, APIcast will make a token request including the generated client assertion as the value of the `client_assertion` parameter.
```

A couple of edits added.

> * `private_key_jwt`: using asymmetric key to request token information from OIDC provider. Prior to a token information request, APIcast will prepare a new JWT authentication token and sign with the key provided. APIcast will then make a token request including the generated client assertion as the value of the `client_assertion` parameter.

Reviewer: ```suggestion
* `private_key_jwt`: When using asymmetric key to request token information from the OIDC provider. Prior to a token information request, APIcast will prepare a new JWT authentication token and sign-in with the key provided. Then, APIcast will make a token request including the generated client assertion as the value of the `client_assertion` parameter.
```

A couple of edits added.

 The response of the Token Introspection Endpoint contains the active attribute. APIcast checks the value of this attribute. Depending on the value of the attribute, APIcast authorizes or rejects the call:

> * `true`: The call is authorized

Reviewer: ```suggestion
* `true`: The call is authorized.
```

Litle typo.

> * `false`: The call is rejected with the Authentication Failed error

Reviewer: ```suggestion
* `false`: The call is rejected with the Authentication Failed error
```

Little typo.

> The policy allows enabling caching of the tokens to avoid calling the Token Introspection Endpoint on every call for the same JWT token. To enable token caching for the Token Introspection Policy, set the max_cached_tokens field to a value from 0, which disables the feature, and 10000. Additionally, you can set a Time to Live (TTL) value from 1 to 3600 seconds for tokens in the max_ttl_tokens field. 

Reviewer: ```suggestion
The policy enables caching tokens to avoid calling the Token Introspection Endpoint on every call for the same JWT token. To enable token caching for the Token Introspection Policy, set the `max_cached_tokens` field to a value between `0`, which disables the feature, and `10000`. Additionally, you can set a Time to Live (TTL) value from `1` to `3600` seconds for tokens in the `max_ttl_tokens` field. 
```

A couple of minor edits.
